diff -Nur a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2018-11-02 14:29:14.933699658 +0800
+++ b/arch/arm/boot/dts/Makefile	2018-11-02 04:13:57.718354065 +0800
@@ -924,6 +924,7 @@
 	sun8i-h3-nanopi-m1-plus.dtb \
 	sun8i-h3-nanopi-neo.dtb \
 	sun8i-h3-nanopi-neo-air.dtb \
+	sun8i-h3-nanopi-neo-oled-spi.dtb \
 	sun8i-h3-orangepi-2.dtb \
 	sun8i-h3-orangepi-lite.dtb \
 	sun8i-h3-orangepi-one.dtb \
diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi
--- a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-02 14:29:14.933699658 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-02 14:39:23.522215453 +0800
@@ -51,6 +51,13 @@
 / {
 	aliases {
 		serial0 = &uart0;
+		i2c0 = &i2c0;
+		i2c1 = &i2c1;
+		i2c2 = &i2c2;
+		spi0 = &spi0;
+		spi1 = &spi1;
+		spiflash = &spiflash;
+		ethernet0 = &emac;
 	};
 
 	chosen {
@@ -65,13 +72,14 @@
 		status {
 			label = "nanopi:blue:status";
 			gpios = <&pio 0 10 GPIO_ACTIVE_HIGH>;
-			linux,default-trigger = "heartbeat";
+			/* linux,default-trigger = "heartbeat"; */
+			default-state = "off";
 		};
 
 		pwr {
 			label = "nanopi:green:pwr";
 			gpios = <&r_pio 0 10 GPIO_ACTIVE_HIGH>;
-			default-state = "on";
+			default-state = "off";
 		};
 	};
 
@@ -89,10 +97,38 @@
 	};
 };
 
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
 &ehci3 {
 	status = "okay";
 };
 
+&ohci3 {
+	status = "okay";
+};
+
 &mmc0 {
 	bus-width = <4>;
 	cd-gpios = <&pio 5 6 GPIO_ACTIVE_HIGH>;
@@ -112,6 +148,11 @@
 		pins = "PA10";
 		function = "gpio_out";
 	};
+
+	spi0_cs_pins: spi0_cs_pins {
+		pins = "PC3", "PA6";
+		function = "gpio_out";
+	};
 };
 
 &r_pio {
@@ -135,3 +176,58 @@
 &usbphy {
 	status = "okay";
 };
+
+&spi0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
+	cs-gpios = <&pio 2 3 GPIO_ACTIVE_HIGH>, <&pio 0 6 GPIO_ACTIVE_HIGH>;
+
+	spiflash: spiflash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0>;
+		status = "disabled";
+
+		spi-max-frequency = <50000000>;
+		mode = <0>;
+	};
+};
+
+&spi1 {
+	// against uart3_rts_cts_pins, so disable
+	status = "disable";
+};
+
+&i2c0 {
+	status = "okay";
+
+	rtc@68 {
+		compatible = "dallas,ds1307";
+		reg = <0x68>;
+	};
+};
+
+&i2c1 {
+    status = "okay";
+};
+
+&i2c2 {
+    status = "okay";
+};
+
+&emac {
+	local-mac-address = [ 00 00 00 00 00 00 ];	
+};
+
+&codec {
+	allwinner,audio-routing =
+		"Line Out", "LINEOUT",
+		"MIC1", "Mic",
+		"Mic",  "MBIAS";
+	status = "okay";
+};
+
diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts
--- a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-02 14:29:14.933699658 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-02 13:51:26.047239730 +0800
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2016 James Pettigrew <james@innovum.com.au>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This file is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This file is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "sun8i-h3-nanopi.dtsi"
+
+/ {
+	model = "FriendlyElec NanoPi NEO";
+	compatible = "friendlyarm,nanopi-neo", "allwinner,sun8i-h3";
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		lbk_fonts: lbk_fonts@0 {
+			reg = <0x4b060000 0x19fc00>;
+		};
+
+		oled_ssd1306_buffer: oled-ssd1306@3c {
+			reg = <0x0 0x0>;
+		};
+	};
+
+	/delete-node/ r_gpio_keys;
+	oled_hat_keys@0 {
+		compatible = "gpio-keys";
+		autorepeat;
+		pinctrl-names = "default";
+		pinctrl-0 = <&oled_hat_pins>;
+		status = "okay";
+
+		sw1@0 {
+			label = "k1";
+			linux,code = <KEY_LEFT>;
+			gpios = <&pio 0 0 GPIO_ACTIVE_HIGH>;
+		};
+
+		sw2@1 {
+			label = "k2";
+			linux,code = <KEY_HOME>;
+			gpios = <&pio 0 2 GPIO_ACTIVE_HIGH>;
+		};
+
+		sw3@2 {
+			label = "k3";
+			linux,code = <KEY_RIGHT>;
+			gpios = <&pio 0 3 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+&reg_usb0_vbus {
+	gpio = <&r_pio 0 2 GPIO_ACTIVE_HIGH>; /* PL2 */
+	status = "okay";
+};
+
+&usb_otg {
+	dr_mode = "otg";
+	status = "okay";
+};
+
+&usbphy {
+	usb0_id_det-gpios = <&pio 6 12 GPIO_ACTIVE_HIGH>; /* PG12 */
+	usb0_vbus-supply = <&reg_usb0_vbus>;
+	status = "okay";
+};
+
+&spiflash {
+    compatible = "jedec,spi-nor";
+    spi-max-frequency = <24000000>;
+    status = "okay";
+
+    partition@0 {
+        reg = <0x0 0x00800000>;
+        label = "whole";
+    };
+
+    partition@1 {
+        reg = <0x0 0x70000>;
+        label = "uboot";
+    };
+
+    partition@2 {
+        reg = <0x70000 0x4000>;
+        label = "uboot_env";
+    };
+
+    partition@3 {
+        reg = <0x74000 0x1000>;
+        label = "openwrt_scr";
+    };
+
+    partition@4 {
+        reg = <0x75000 0x6000>;
+        label = "dtb";
+    };
+
+    partition@5 {
+        reg = <0x7b000 0x60000>;
+        label = "font";
+    };
+
+    partition@6 {
+        reg = <0xdb000 0x2e0000>;
+        label = "kernel";
+    };
+
+    partition@7 {
+        reg = <0x3bb000 0x445000>;
+        label = "rootfs";
+    };
+};
+
+&i2c0 {
+    clock-frequency = <400000>;
+    status = "okay";
+
+    oled-ssd1306@3c {
+        compatible = "soloman,ssd1306-i2c";
+        reg = <0x3c>;
+        width = <128>;
+        height = <64>;
+        flip_vertical = <0>;
+        buffer = <&oled_ssd1306_buffer>;
+    };
+};
+
+&uart2 {
+	status = "disabled"; /* for OLED Hat Keys */
+};
+
+&pio {
+	oled_hat_pins: oled_hat_pins@0 {
+		pins = "PA0", "PA2", "PA3";
+		function = "gpio_in";
+	};
+};
+
+&r_pio {
+	/delete-node/ key_pins@0;
+};
+
diff -Nur a/arch/arm/boot/dts/sunxi-h3-h5.dtsi b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
--- a/arch/arm/boot/dts/sunxi-h3-h5.dtsi	2018-11-02 04:13:57.698354202 +0800
+++ b/arch/arm/boot/dts/sunxi-h3-h5.dtsi	2018-11-02 14:40:21.479332786 +0800
@@ -593,7 +593,7 @@
 
 		i2c2: i2c@01c2b400 {
 			compatible = "allwinner,sun6i-a31-i2c";
-			reg = <0x01c2b000 0x400>;
+			reg = <0x01c2b400 0x400>;
 			interrupts = <GIC_SPI 8 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&ccu CLK_BUS_I2C2>;
 			resets = <&ccu RST_BUS_I2C2>;
@@ -647,6 +647,20 @@
 			status = "disabled";
 		};
 
+		r_i2c: i2c@01f02400 {
+			compatible = "allwinner,sun6i-a31-i2c";
+			reg = <0x01f02400 0x400>;
+			interrupts = <GIC_SPI 44 IRQ_TYPE_LEVEL_HIGH>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&r_i2c_pins_a>;
+			clocks = <&r_ccu CLK_APB0_I2C>;
+			clock-frequency = <100000>;
+			resets = <&r_ccu RST_APB0_I2C>;
+			status = "disabled";
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
 		r_pio: pinctrl@01f02c00 {
 			compatible = "allwinner,sun8i-h3-r-pinctrl";
 			reg = <0x01f02c00 0x400>;
@@ -662,6 +676,11 @@
 				pins = "PL11";
 				function = "s_cir_rx";
 			};
+
+			r_i2c_pins_a: r_i2c@0 {
+				pins = "PL0", "PL1";
+				function = "s_i2c";
+			};
 		};
 	};
 };
