diff -Nur a/arch/arm/boot/dts/sunxi-h3-h5.dtsi b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
--- a/arch/arm/boot/dts/sunxi-h3-h5.dtsi	2018-11-23 10:36:25.579068722 +0800
+++ b/arch/arm/boot/dts/sunxi-h3-h5.dtsi	2018-11-23 10:54:56.788815690 +0800
@@ -295,6 +295,11 @@
 				drive-strength = <40>;
 			};
 
+			i2s0_pins: i2s0 {
+				pins = "PA18", "PA19", "PA20", "PA21";
+				function = "i2s0";
+			};
+
 			i2c0_pins: i2c0 {
 				pins = "PA11", "PA12";
 				function = "i2c0";
@@ -525,6 +530,23 @@
 			status = "disabled";
 		};
 
+		i2s0: i2s@01c22000 {
+ 			#sound-dai-cells = <0>;
+ 			compatible = "allwinner,sun8i-h3-i2s";
+ 			reg = <0x01c22000 0x400>;
+ 			interrupts = <GIC_SPI 13 IRQ_TYPE_LEVEL_HIGH>; /* 45-32=13, P207 */
+ 			clocks = <&ccu CLK_BUS_I2S0>, <&ccu CLK_I2S0>;
+ 			clock-names = "apb", "mod";
+ 			resets = <&ccu RST_BUS_I2S0>;
+ 			pinctrl-names = "default";
+ 			pinctrl-0 = <&i2s0_pins>;
+ 			dmas = <&dma 3>, <&dma 3>;
+ 			dma-names = "rx", "tx";
+ 			status = "disabled";
+ 			#address-cells = <1>;
+ 			#size-cells = <0>;
+ 		};
+
 		uart0: serial@01c28000 {
 			compatible = "snps,dw-apb-uart";
 			reg = <0x01c28000 0x400>;
diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi
--- a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-23 10:36:25.579068722 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-23 10:58:35.762563118 +0800
@@ -60,6 +60,8 @@
 		spi1 = &spi1;
 		spiflash = &spiflash;
 		ethernet0 = &emac;
+		i2s0 = &i2s0;
+		pcm5102a = &pcm5102a;
 	};
 
 	chosen {
@@ -117,6 +119,12 @@
 		states = <1100000 0x0
 			  1300000 0x1>;
 	};
+
+	pcm5102a: pcm5102a-codec {
+		#sound-dai-cells = <0>;
+		compatible = "ti,pcm5102a";
+		status = "disabled";
+	};
 };
 
 &cpu0 {
@@ -316,3 +324,8 @@
 	status = "okay";
 };
 
+&i2s0 {
+    sound-dai = <&pcm5102a>;
+    status = "disabled";
+};
+
diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts
--- a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-23 10:36:25.567069306 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-23 11:05:08.663200395 +0800
@@ -171,6 +171,10 @@
     };
 };
 
+&i2c1 {
+	status = "disabled"; /* for I2S0 */
+};
+
 &uart2 {
 	status = "disabled"; /* for OLED Hat Keys */
 };
diff -Nur a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts	2018-11-23 10:36:25.587068332 +0800
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts	2018-11-23 11:12:12.458300788 +0800
@@ -55,9 +55,11 @@
 	aliases {
 		serial0 = &uart0;
 		i2c0 = &i2c0;
+		i2c1 = &i2c1;
 		spi0 = &spi0;
 		spiflash = &spiflash;
 		ethernet0 = &emac;
+		i2s0 = &i2s0;
 	};
 
 	reserved-memory {
@@ -167,8 +169,13 @@
 		states = <1100000 0x0
 			  1100000 0x1>;
 	};
-};
 
+	pcm5102a: pcm5102a-codec {
+		#sound-dai-cells = <0>;
+		compatible = "ti,pcm5102a";
+		status = "disabled";
+	};
+};
 
 &cpu0 {
 	operating-points = <
@@ -234,6 +241,11 @@
 	status = "okay";
 };
 
+&i2s0 {
+    sound-dai = <&pcm5102a>;
+    status = "disabled";
+};
+
 &ehci0 {
 	status = "okay";
 };
@@ -304,6 +316,10 @@
 	status = "disabled";
 };
 
+&i2c1 {
+	status = "disabled"; /* for I2S0 */
+};
+
 &uart2 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart2_pins>;
@@ -322,7 +338,6 @@
 	status = "okay";
 };
 
-
 &spi0 {
 	#address-cells = <1>;
 	#size-cells = <0>;
