diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi
--- a/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-26 03:05:34.561200140 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi.dtsi	2018-11-27 01:25:50.099737095 +0800
@@ -58,6 +58,8 @@
 		i2c2 = &i2c2;
 		spi0 = &spi0;
 		spi1 = &spi1;
+		spidev0 = &spidev0;
+		spidev1 = &spidev1;
 		spiflash = &spiflash;
 		ethernet0 = &emac;
 		i2s0 = &i2s0;
@@ -235,9 +237,15 @@
 		function = "gpio_out";
 	};
 
+	spi0_pins: spi0 {
+		pins = "PC0", "PC1", "PC2";
+		function = "spi0";
+	};
+
 	spi0_cs_pins: spi0_cs_pins {
 		pins = "PC3", "PA6";
 		function = "gpio_out";
+		bias-disable;
 	};
 };
 
@@ -279,6 +287,22 @@
 	pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
 	cs-gpios = <&pio 2 3 GPIO_ACTIVE_HIGH>, <&pio 0 6 GPIO_ACTIVE_HIGH>;
 
+	spidev0: spi@0 {
+		compatible = "nanopi,spidev";
+		reg = <0>;
+		status = "disabled";
+
+		spi-max-frequency = <10000000>;
+	};
+
+	spidev1: spi@1 {
+		compatible = "nanopi,spidev";
+		reg = <1>;
+		status = "disabled";
+
+		spi-max-frequency = <10000000>;
+	};
+
 	spiflash: spiflash@0 {
 		#address-cells = <1>;
 		#size-cells = <1>;
diff -Nur a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts
--- a/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-26 03:05:34.561200140 +0800
+++ b/arch/arm/boot/dts/sun8i-h3-nanopi-neo-oled-spi.dts	2018-11-27 01:27:00.737405848 +0800
@@ -111,6 +111,11 @@
 	status = "okay";
 };
 
+&spidev1 {
+    /* use PA6 as CS pin */
+    status = "okay";
+};
+
 &spiflash {
     compatible = "jedec,spi-nor";
     spi-max-frequency = <24000000>;
diff -Nur a/drivers/spi/spidev.c b/drivers/spi/spidev.c
--- a/drivers/spi/spidev.c     2018-11-26 03:05:33.413213691 +0800
+++ b/drivers/spi/spidev.c     2018-11-27 01:27:37.370193969 +0800
@@ -670,6 +670,7 @@
 	{ .compatible = "ge,achc" },
 	{ .compatible = "semtech,sx1301" },
 	{ .compatible = "siliconlabs,si3210" },
+	{ .compatible = "nanopi,spidev" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, spidev_dt_ids);
diff -Nur a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts	2018-11-27 01:20:45.617570233 +0800
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo2-oled-spi.dts	2018-11-27 01:47:07.265575178 +0800
@@ -57,6 +57,7 @@
 		i2c0 = &i2c0;
 		i2c1 = &i2c1;
 		spi0 = &spi0;
+		spidev1 = &spidev1;
 		spiflash = &spiflash;
 		ethernet0 = &emac;
 		i2s0 = &i2s0;
@@ -347,6 +348,15 @@
 	pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
 	cs-gpios = <&pio 2 3 GPIO_ACTIVE_HIGH>, <&pio 0 6 GPIO_ACTIVE_HIGH>;
 
+	spidev1: spi@1 {
+		/* use PA6 as CS pin */
+		compatible = "nanopi,spidev";
+		reg = <1>;
+		status = "okay";
+
+		spi-max-frequency = <10000000>;
+	};
+
 	spiflash: spiflash@0 {
 		#address-cells = <1>;
 		#size-cells = <1>;
@@ -418,9 +428,15 @@
 		function = "gpio_in";
 	};
 
+	spi0_pins: spi0 {
+		pins = "PC0", "PC1", "PC2";
+		function = "spi0";
+	};
+
 	spi0_cs_pins: spi0_cs_pins {
 		pins = "PC3", "PA6";
 		function = "gpio_out";
+		bias-disable;
 	};
 };
 
diff -Nur a/drivers/spi/spi-sun6i.c b/drivers/spi/spi-sun6i.c
--- a/drivers/spi/spi-sun6i.c	2018-11-14 03:15:18.000000000 +0800
+++ b/drivers/spi/spi-sun6i.c	2018-11-29 17:52:23.202911410 +0800
@@ -21,6 +21,7 @@
 #include <linux/platform_device.h>
 #include <linux/pm_runtime.h>
 #include <linux/reset.h>
+#include <linux/gpio.h>
 
 #include <linux/spi/spi.h>
 
@@ -180,16 +181,22 @@
 	struct sun6i_spi *sspi = spi_master_get_devdata(spi->master);
 	u32 reg;
 
+	enable = !enable; // ODD!!! see driver/spi/spi.c::spi_set_cs()
+
 	reg = sun6i_spi_read(sspi, SUN6I_TFR_CTL_REG);
 	reg &= ~SUN6I_TFR_CTL_CS_MASK;
 	reg |= SUN6I_TFR_CTL_CS(spi->chip_select);
 
 	if (enable)
-		reg |= SUN6I_TFR_CTL_CS_LEVEL;
-	else
 		reg &= ~SUN6I_TFR_CTL_CS_LEVEL;
+	else
+		reg |= SUN6I_TFR_CTL_CS_LEVEL;
 
 	sun6i_spi_write(sspi, SUN6I_TFR_CTL_REG, reg);
+
+	if (spi->cs_gpio >= 0) {
+		ndelay(150);
+	}
 }
 
 static size_t sun6i_spi_max_transfer_size(struct spi_device *spi)
@@ -440,7 +447,7 @@
 	struct spi_master *master;
 	struct sun6i_spi *sspi;
 	struct resource	*res;
-	int ret = 0, irq;
+	int ret = 0, irq, i;
 
 	master = spi_alloc_master(&pdev->dev, sizeof(struct sun6i_spi));
 	if (!master) {
@@ -485,6 +492,7 @@
 	master->dev.of_node = pdev->dev.of_node;
 	master->auto_runtime_pm = true;
 	master->max_transfer_size = sun6i_spi_max_transfer_size;
+	master->flags = SPI_MASTER_GPIO_SS;
 
 	sspi->hclk = devm_clk_get(&pdev->dev, "ahb");
 	if (IS_ERR(sspi->hclk)) {
@@ -529,6 +537,15 @@
 		goto err_pm_disable;
 	}
 
+	if (master->cs_gpios) {
+		for (i = 0; i < master->num_chipselect; i++) {
+			if (master->cs_gpios[i] >= 0) {
+				gpio_set_value(master->cs_gpios[i], 1);
+				ndelay(150);
+			}
+		}
+	}
+
 	return 0;
 
 err_pm_disable:
