diff -Nur a/drivers/rtc/Kconfig b/drivers/rtc/Kconfig
--- a/drivers/rtc/Kconfig	2020-02-03 12:49:18.044178210 +0800
+++ b/drivers/rtc/Kconfig	2020-02-03 13:04:42.296967477 +0800
@@ -1461,7 +1461,7 @@
 	  based RTC on SUN4V systems.
 
 config RTC_DRV_SUN6I
-	bool "Allwinner A31 RTC"
+	tristate "Allwinner A31 RTC"
 	default MACH_SUN6I || MACH_SUN8I
 	depends on COMMON_CLK
 	depends on ARCH_SUNXI || COMPILE_TEST
